driver-core/src/main/java/com/datastax/driver/core/RegularStatement.java

```diff
@@ -42,32 +70,1143 @@ protected RegularStatement() {}
+         *
+         * @return the index, or -1 if named variables were used on this statement.
+         */
+        public int getIndex() {
+            return index;
+        }
+
+        /**
+         * Return the name of the value in the statement.
+         * <p>
+         * This information is set if a named setter (for example {@link #setDouble(String, double)}) was used to set the value.
+         * Note that the variables of a given statement are either all named or all positional.
+         *
+         * @return the name, or {@code null} if positional variables were used on this statement.
+         */
+        public String getName() {
+            return name;
+        }
+
+        /**
+         * Return the CQL data type inferred for the value.
+         * <p>
+         * This information is determined from the method that was used to set the value. For example, if
+         * {@link #setDouble(String, double)} was used, it will be {@link DataType#cdouble()}. For non-standard types, it will
+         * be the CQL type of the first {@link CodecRegistry custom codec} that can handle the value.
+         * The only case where the CQL type can't be inferred is if the value was set with {@link #setBytesUnsafe(int, ByteBuffer)}
+         * or {@link #setToNull(int)}; then this method will return {@code null}.
+         *
+         * @return the data type, or {@code null} if this variable was set with {@link #setBytesUnsafe(int, ByteBuffer)}
+         * or {@link #setToNull(int)}.
+         */
+        public DataType getType() {
+            return type;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s:%s",
+                (index >= 0) ? index : name,
+                (type == null) ? "<unknown type>" : type.toString());
+        }
+    }
+
+    // A value that was set on the statement.
+    private class Value {
+
+        private final ValueDefinition definition;
+
+        // The serialized form that will be sent to Cassandra alongside the query
+        private final ByteBuffer bytes;
+
+        private Value(ValueDefinition definition, ByteBuffer bytes) {
+            this.definition = definition;
+            this.bytes = bytes;
+        }
+
+        private <V> V as(TypeToken<V> javaType) {
+            if (OBJECT_TYPE.equals(javaType)) { // we're handling a getObject() call
+                @SuppressWarnings("unchecked")
+                V v = (V)asObject(); // Since V is Object, this is a safe cast
+                return v;
+            } else { // we're handling a typed getter call
+                assert javaType != null;
+                return with(codecFor(javaType));
+            }
+        }
+
+        private <V> V with(TypeCodec<V> codec) {
+            return codec.deserialize(bytes, protocolVersion);
+        }
+
+        private boolean isNull() {
+            return bytes == null || bytes.remaining() == 0;
```

Comment:
That's not necessarily true for all CQL types, e.g. for varchars, `bytes == null`-> `null` but `bytes.remaining() == 0` -> empty string. I think it would be safer to store a flag to tell if the value was null or not, or better yet, store the original java type let the codec decide: `return  with(codecFor(javaType)) == null`. Let's keep in mind that a codec can choose never to return nulls, e.g. `OptionalCodec`.